crates/supervisor/storage/src/providers/log_provider.rs

```diff
@@ -142,23 +142,110 @@ where
     /// Rewinds the log storage by deleting all blocks and logs from the given block onward.
     /// Fails if the given block exists with a mismatching hash (to prevent unsafe deletion).
     pub(crate) fn rewind_to(&self, block: &BlockNumHash) -> Result<(), StorageError> {
-        let mut cursor = self.tx.cursor_write::<BlockRefs>()?;
-        let mut walker = cursor.walk(Some(block.number))?;
+        info!(
+            target: "supervisor::storage",
+            chain_id = %self.chain_id,
+            target_block_number = %block.number,
+            target_block_hash = %block.hash,
+            "Starting rewind of log storage"
+        );
 
-        while let Some(Ok((key, stored_block))) = walker.next() {
-            if key == block.number && block.hash != stored_block.hash {
-                error!(
+        // Get the latest block number from BlockRefs
+        let latest_block = {
+            let mut cursor = self.tx.cursor_read::<BlockRefs>()?;
+            cursor.last()?.map(|(num, _)| num).unwrap_or(block.number)
+        };
+
+        // Check for future block
+        if block.number > latest_block {
+            error!(
+                target: "supervisor::storage",
+                chain_id = %self.chain_id,
+                target_block_number = %block.number,
+                latest_block,
+                "Cannot rewind to future block"
+            );
+            return Err(StorageError::FutureData);
+        }
+
+        let total_blocks = latest_block.saturating_sub(block.number);
+        const LOG_INTERVAL: u64 = 100;
+        let mut processed_blocks = 0;
+
+        // Delete all blocks and logs with number ≥ `block.number`
+        {
+            let mut cursor = self.tx.cursor_write::<BlockRefs>()?;
+            let mut walker = cursor.walk(Some(block.number))?;
+
+            info!(
+                target: "supervisor::storage",
+                chain_id = %self.chain_id,
+                target_block_number = %block.number,
+                target_block_hash = %block.hash,
+                latest_block,
+                total_blocks,
+                "Starting rewind of block storage"
+            );
+
+            while let Some(Ok((key, stored_block))) = walker.next() {
+                if key == block.number && block.hash != stored_block.hash {
+                    error!(
+                        target: "supervisor::storage",
+                        chain_id = %self.chain_id,
+                        %stored_block,
+                        incoming_block = ?block,
+                        "Requested block to rewind does not match stored block"
+                    );
+                    return Err(StorageError::ConflictError);
```

Comment:
aware that this is out of scope of this pr, but this should probably rather be `SuperchainDAError::DataCorruption` @dhyaniarun1993